diff --git a/src/components/tree/phyloTree/grid.js b/src/components/tree/phyloTree/grid.js
index ab17e1a750ff08575d477a5f7e042fa2134f5d3d..5d7f781cc7037fa66c8b71ae8b3ab59aee6417fd 100644
--- a/src/components/tree/phyloTree/grid.js
+++ b/src/components/tree/phyloTree/grid.js
@@ -119,7 +119,7 @@ const computeNumericGridPoints = (minVal, maxVal, layout, nMinorTicks, axis) =>
 const calculateTemporalGridSeperation = (timeRange, pxAvailable) => {
   const [majorStep, minorStep] = [{unit: "DAY", n: 1}, {unit: "DAY", n: 0}];
   const minPxBetweenMajorGrid = (pxAvailable < 1000 ? 130 : 180);
-  const timeBetweenMajorGrids = timeRange/(Math.floor(pxAvailable / minPxBetweenMajorGrid));
+  const timeBetweenMajorGrids = timeRange/Math.max(1, Math.floor(pxAvailable / minPxBetweenMajorGrid));
   const levels = {
     CENTURY: {t: 100, max: undefined},
     DECADE: {t: 10, max: 5}, // i.e. spacing of 50 years is ok, but 60 jumps up to 100y spacing
